core/src/main/scala/org/allenai/pipeline/ArtifactFactory.scala

```diff
@@ -0,0 +1,126 @@
+package org.allenai.pipeline
+
+import scala.reflect.ClassTag
+
+import java.io.File
+import java.net.URI
+
+/** Creates an Artifact from a URL
+  */
+trait ArtifactFactory {
+  /** @param url The location of the Artifact.  The scheme (protocol) is used to determine the
+    *            specific implementation.
+    * @tparam A The type of the Artifact to create.  May be an abstract or concrete type
+    * @return The artifact
+    */
+  def createArtifact[A <: Artifact: ClassTag](url: URI): A
```

Comment:
This is an extensible API for creating Artifacts from URLs.  Previously, the `Artifact` class had a `url` method, but there was no way to go in the other direction.